File: c/palindrome_linked_list.c
#include <stdbool.h>
#include <stdlib.h>

struct ListNode {
	int              val;
	struct ListNode *next;
};

struct ListNode *
reverse(struct ListNode *head) {
	struct ListNode *curr = head, *prev = NULL, *next = NULL;
	while (curr) {
		next       = curr->next;
		curr->next = prev;
		prev       = curr;
		curr       = next;
	}
	return prev;
}

bool
equal(struct ListNode *l, struct ListNode *m) {
	while (l && m) {
		if (l->val != m->val)
			return false;
		l = l->next;
		m = m->next;
	}
	return true;
}

bool
isPalindrome(struct ListNode *head) {
	struct ListNode *slow = head, *fast = head;
	while (fast && fast->next) {
		slow = slow->next;
		fast = fast->next->next;
	}
	return equal(head, reverse(slow));
}

File: go/palindrome_linked_list.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	vals := []int{}
	for node := head; node != nil; node = node.Next {
		vals = append(vals, node.Val)
	}
	for l, r := 0, len(vals)-1; l < r; l, r = l+1, r-1 {
		if vals[l] != vals[r] {
			return false
		}
	}
	return true
}
